File: inspect_live_page.py
#!/usr/bin/env python3
"""
Inspect live page at http://127.0.0.1:XXXX/team_capacity_planner.html
Uses playwright (installed via pip) or requests+selenium as fallback.
"""
import sys
import time
import json
import subprocess
import re

try:
    from playwright.sync_api import sync_playwright
    HAS_PLAYWRIGHT = True
except ImportError:
    HAS_PLAYWRIGHT = False
    print("⚠️  Playwright not available, trying alternative methods...")

def inspect_with_playwright():
    """Use Playwright to inspect the page."""
    ports = [3000, 3001, 4173, 5000, 8000, 8080]
    
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context()
        page = context.new_page()
        
        console_logs = []
        page_errors = []
        network_errors = []
        
        def handle_console(msg):
            console_logs.append({
                'type': msg.type,
                'text': msg.text
            })
        
        def handle_error(err):
            page_errors.append({
                'message': str(err)
            })
        
        page.on("console", handle_console)
        page.on("pageerror", handle_error)
        
        url = None
        for port in ports:
            try:
                test_url = f"http://127.0.0.1:{port}/team_capacity_planner.html"
                try:
                    response = page.goto(test_url, wait_until="networkidle", timeout=5000)
                    if response and response.ok:
                        url = test_url
                        print(f"✓ Connected to {url}\n")
                        break
                except:
                    pass
            except:
                pass
        
        if not url:
            print("✗ Could not connect to any port (3000, 3001, 4173, 5000, 8000, 8080)")
            browser.close()
            return False
        
        # Wait for page to load
        page.wait_for_timeout(2000)
        
        # Evaluate page structure
        findings = page.evaluate("""() => {
            const result = {
                wiContainerExists: !!document.getElementById('wi-container'),
                epicRowsCount: 0,
                firstRowsInfo: [],
                epicRowStyles: {},
                epicRowHeadStyles: {},
                boundingBoxes: {}
            };
            
            const wiContainer = document.getElementById('wi-container');
            if (wiContainer) {
                const epicRows = wiContainer.querySelectorAll('.epic-row');
                result.epicRowsCount = epicRows.length;
                
                for (let i = 0; i < Math.min(3, epicRows.length); i++) {
                    const row = epicRows[i];
                    const rowHead = row.querySelector('.epic-row-head');
                    const rowText = row.textContent || '';
                    const headText = rowHead ? (rowHead.textContent || '') : '';
                    
                    result.firstRowsInfo.push({
                        rowIndex: i,
                        textContent: rowText.substring(0, 150),
                        headTextContent: headText.substring(0, 100),
                        rowHeight: row.offsetHeight,
                        rowWidth: row.offsetWidth,
                        headHeight: rowHead ? rowHead.offsetHeight : null,
                        display: window.getComputedStyle(row).display,
                        visibility: window.getComputedStyle(row).visibility,
                        height: window.getComputedStyle(row).height,
                        minHeight: window.getComputedStyle(row).minHeight,
                        maxHeight: window.getComputedStyle(row).maxHeight,
                        overflow: window.getComputedStyle(row).overflow
                    });
                }
                
                if (epicRows.length > 0) {
                    const firstRow = epicRows[0];
                    const styles = window.getComputedStyle(firstRow);
                    result.epicRowStyles = {
                        display: styles.display,
                        visibility: styles.visibility,
                        height: styles.height,
                        minHeight: styles.minHeight,
                        maxHeight: styles.maxHeight,
                        overflow: styles.overflow,
                        opacity: styles.opacity,
                        lineHeight: styles.lineHeight,
                        fontSize: styles.fontSize,
                        backgroundColor: styles.backgroundColor
                    };
                    
                    result.boundingBoxes.epicRow = {
                        width: firstRow.getBoundingClientRect().width,
                        height: firstRow.getBoundingClientRect().height
                    };
                    
                    const rowHead = firstRow.querySelector('.epic-row-head');
                    if (rowHead) {
                        const headStyles = window.getComputedStyle(rowHead);
                        result.epicRowHeadStyles = {
                            display: headStyles.display,
                            height: headStyles.height,
                            minHeight: headStyles.minHeight,
                            maxHeight: headStyles.maxHeight,
                            overflow: headStyles.overflow,
                            visibility: headStyles.visibility,
                            opacity: headStyles.opacity,
                            fontSize: headStyles.fontSize,
                            lineHeight: headStyles.lineHeight
                        };
                        
                        result.boundingBoxes.epicRowHead = {
                            width: rowHead.getBoundingClientRect().width,
                            height: rowHead.getBoundingClientRect().height
                        };
                    }
                }
            }
            
            return result;
        }""")
        
        # Print findings
        print("=" * 50)
        print("PAGE INSPECTION FINDINGS")
        print("=" * 50)
        
        print("\n1. WI-CONTAINER & EPIC-ROWS:")
        print(f"   - #wi-container exists: {findings['wiContainerExists']}")
        print(f"   - .epic-row count: {findings['epicRowsCount']}")
        print(f"   - Contains epic-row nodes: {'YES' if findings['epicRowsCount'] > 0 else 'NO'}")
        
        if findings['firstRowsInfo']:
            print("\n2. FIRST 1-3 ROWS CONTENT:")
            for row in findings['firstRowsInfo']:
                print(f"\n   Row {row['rowIndex']}:")
                text = row['textContent']
                if text:
                    print(f"   - textContent: \"{text[:80]}{'...' if len(text) > 80 else ''}\"")
                head_text = row['headTextContent']
                if head_text:
                    print(f"   - headTextContent: \"{head_text[:80]}{'...' if len(head_text) > 80 else ''}\"")
                print(f"   - Dimensions: {row['rowWidth']}x{row['rowHeight']}px (head height: {row['headHeight']}px)")
        
        print("\n3. COMPUTED STYLES FOR .epic-row:")
        for key, val in findings['epicRowStyles'].items():
            if val and val not in ['auto', 'none', '']:
                print(f"   - {key}: {val}")
        
        print("\n4. COMPUTED STYLES FOR .epic-row-head:")
        for key, val in findings['epicRowHeadStyles'].items():
            if val and val not in ['auto', 'none', '']:
                print(f"   - {key}: {val}")
        
        print("\n5. BOUNDING BOXES:")
        if findings['boundingBoxes'].get('epicRow'):
            bb = findings['boundingBoxes']['epicRow']
            print(f"   - .epic-row: {bb['width']:.0f}x{bb['height']:.0f}px")
        if findings['boundingBoxes'].get('epicRowHead'):
            bb = findings['boundingBoxes']['epicRowHead']
            print(f"   - .epic-row-head: {bb['width']:.0f}x{bb['height']:.0f}px")
        
        print("\n6. CONSOLE ERRORS/WARNINGS:")
        errors = [m for m in console_logs if m['type'] in ['error', 'warning']]
        if errors:
            for msg in errors[:5]:
                print(f"   [{msg['type']}] {msg['text']}")
        else:
            print("   (none)")
        
        print("\n7. PAGE ERRORS:")
        if page_errors:
            for err in page_errors[:5]:
                print(f"   {err['message']}")
        else:
            print("   (none)")
        
        browser.close()
        return True


if __name__ == '__main__':
    if HAS_PLAYWRIGHT:
        success = inspect_with_playwright()
        sys.exit(0 if success else 1)
    else:
        print("ERROR: Playwright is required but not installed.")
        print("Install with: pip install playwright")
        sys.exit(1)

File: inspect_live_page.js
const { chromium } = require('playwright');

(async () => {
  const browser = await chromium.launch({ headless: true });
  const context = await browser.createContext();
  const page = await context.newPage();

  // Listen for console messages
  const consoleLogs = [];
  page.on('console', msg => {
    consoleLogs.push({
      type: msg.type(),
      text: msg.text(),
      location: msg.location()
    });
  });

  // Listen for page errors
  const pageErrors = [];
  page.on('pageerror', err => {
    pageErrors.push({
      message: err.message,
      stack: err.stack
    });
  });

  // Listen for network errors
  const networkErrors = [];
  page.on('requestfailed', req => {
    networkErrors.push({
      url: req.url(),
      error: req.failure().errorText
    });
  });

  const ports = [3000, 3001, 4173, 5000, 8000, 8080];
  let connected = false;
  let url = '';

  for (const port of ports) {
    try {
      url = `http://127.0.0.1:${port}/team_capacity_planner.html`;
      const response = await page.goto(url, { waitUntil: 'networkidle', timeout: 5000 }).catch(() => null);
      if (response && response.ok()) {
        console.log(`✓ Connected to ${url}`);
        connected = true;
        break;
      }
    } catch (e) {
      // Try next port
    }
  }

  if (!connected) {
    console.log('✗ Could not connect to any port');
    await browser.close();
    process.exit(1);
  }

  // Wait a bit for page to fully load
  await page.waitForTimeout(2000);

  // Inspect the page
  const findings = await page.evaluate(() => {
    const result = {
      wiContainerExists: !!document.getElementById('wi-container'),
      epicRowsInContainer: [],
      epicRowsCount: 0,
      firstRowsInfo: [],
      epicRowStyles: {},
      epicRowHeadStyles: {},
      boundingBoxes: {}
    };

    const wiContainer = document.getElementById('wi-container');
    if (wiContainer) {
      const epicRows = wiContainer.querySelectorAll('.epic-row');
      result.epicRowsCount = epicRows.length;

      // Get first 1-3 rows content
      for (let i = 0; i < Math.min(3, epicRows.length); i++) {
        const row = epicRows[i];
        const rowHead = row.querySelector('.epic-row-head');
        result.firstRowsInfo.push({
          rowIndex: i,
          innerHTML: row.innerHTML.substring(0, 200) + (row.innerHTML.length > 200 ? '...' : ''),
          textContent: (row.textContent || '').substring(0, 150),
          classList: Array.from(row.classList),
          headInnerHTML: rowHead ? rowHead.innerHTML.substring(0, 150) : null,
          headTextContent: rowHead ? (rowHead.textContent || '').substring(0, 100) : null,
          rowHeight: row.offsetHeight,
          rowWidth: row.offsetWidth,
          headHeight: rowHead ? rowHead.offsetHeight : null,
          headWidth: rowHead ? rowHead.offsetWidth : null,
          display: window.getComputedStyle(row).display,
          visibility: window.getComputedStyle(row).visibility,
          height: window.getComputedStyle(row).height,
          minHeight: window.getComputedStyle(row).minHeight,
          maxHeight: window.getComputedStyle(row).maxHeight,
          overflow: window.getComputedStyle(row).overflow,
          headDisplay: rowHead ? window.getComputedStyle(rowHead).display : null,
          headHeight_computed: rowHead ? window.getComputedStyle(rowHead).height : null
        });
      }

      // Get computed styles for first epic-row
      if (epicRows.length > 0) {
        const firstRow = epicRows[0];
        const styles = window.getComputedStyle(firstRow);
        result.epicRowStyles = {
          display: styles.display,
          visibility: styles.visibility,
          height: styles.height,
          minHeight: styles.minHeight,
          maxHeight: styles.maxHeight,
          width: styles.width,
          overflow: styles.overflow,
          opacity: styles.opacity,
          pointerEvents: styles.pointerEvents,
          position: styles.position,
          top: styles.top,
          left: styles.left,
          backgroundColor: styles.backgroundColor,
          padding: styles.padding,
          margin: styles.margin,
          borderWidth: styles.borderWidth,
          lineHeight: styles.lineHeight,
          fontSize: styles.fontSize
        };

        // Get bounding boxes
        result.boundingBoxes.epicRow = {
          top: firstRow.getBoundingClientRect().top,
          left: firstRow.getBoundingClientRect().left,
          width: firstRow.getBoundingClientRect().width,
          height: firstRow.getBoundingClientRect().height,
          bottom: firstRow.getBoundingClientRect().bottom,
          right: firstRow.getBoundingClientRect().right
        };

        const rowHead = firstRow.querySelector('.epic-row-head');
        if (rowHead) {
          const headStyles = window.getComputedStyle(rowHead);
          result.epicRowHeadStyles = {
            display: headStyles.display,
            height: headStyles.height,
            minHeight: headStyles.minHeight,
            maxHeight: headStyles.maxHeight,
            width: headStyles.width,
            visibility: headStyles.visibility,
            opacity: headStyles.opacity,
            overflow: headStyles.overflow,
            backgroundColor: headStyles.backgroundColor,
            padding: headStyles.padding,
            margin: headStyles.margin,
            fontSize: headStyles.fontSize,
            lineHeight: headStyles.lineHeight
          };

          result.boundingBoxes.epicRowHead = {
            top: rowHead.getBoundingClientRect().top,
            left: rowHead.getBoundingClientRect().left,
            width: rowHead.getBoundingClientRect().width,
            height: rowHead.getBoundingClientRect().height,
            bottom: rowHead.getBoundingClientRect().bottom,
            right: rowHead.getBoundingClientRect().right
          };
        }
      }
    }

    return result;
  });

  console.log('\n=== PAGE INSPECTION FINDINGS ===\n');
  console.log('1. WI-CONTAINER & EPIC-ROWS:');
  console.log(`   - #wi-container exists: ${findings.wiContainerExists}`);
  console.log(`   - .epic-row count: ${findings.epicRowsCount}`);
  console.log(`   - Contains epic-row nodes: ${findings.epicRowsCount > 0 ? 'YES' : 'NO'}`);

  if (findings.firstRowsInfo.length > 0) {
    console.log('\n2. FIRST 1-3 ROWS CONTENT:');
    findings.firstRowsInfo.forEach((row, idx) => {
      console.log(`\n   Row ${idx}:`);
      console.log(`   - textContent: "${row.textContent.substring(0, 80)}${row.textContent.length > 80 ? '...' : ''}"`);
      console.log(`   - headTextContent: "${row.headTextContent ? row.headTextContent.substring(0, 80) : 'N/A'}"`);
      console.log(`   - Dimensions: ${row.rowWidth}x${row.rowHeight}px (head: ${row.headWidth || 'N/A'}x${row.headHeight || 'N/A'}px)`);
    });
  }

  if (Object.keys(findings.epicRowStyles).length > 0) {
    console.log('\n3. COMPUTED STYLES FOR .epic-row:');
    Object.entries(findings.epicRowStyles).forEach(([key, val]) => {
      if (val !== 'auto' && val !== '' && val !== 'none') {
        console.log(`   - ${key}: ${val}`);
      }
    });
  }

  if (Object.keys(findings.epicRowHeadStyles).length > 0) {
    console.log('\n4. COMPUTED STYLES FOR .epic-row-head:');
    Object.entries(findings.epicRowHeadStyles).forEach(([key, val]) => {
      if (val !== 'auto' && val !== '' && val !== 'none') {
        console.log(`   - ${key}: ${val}`);
      }
    });
  }

  if (findings.boundingBoxes.epicRow) {
    console.log('\n5. BOUNDING BOXES:');
    console.log(`   - .epic-row: width=${findings.boundingBoxes.epicRow.width.toFixed(2)}px, height=${findings.boundingBoxes.epicRow.height.toFixed(2)}px`);
    if (findings.boundingBoxes.epicRowHead) {
      console.log(`   - .epic-row-head: width=${findings.boundingBoxes.epicRowHead.width.toFixed(2)}px, height=${findings.boundingBoxes.epicRowHead.height.toFixed(2)}px`);
    }
  }

  console.log('\n6. CONSOLE ERRORS:');
  if (consoleLogs.length > 0) {
    consoleLogs.slice(0, 10).forEach(log => {
      if (log.type === 'error' || log.type === 'warning') {
        console.log(`   [${log.type}] ${log.text}`);
      }
    });
    if (consoleLogs.filter(l => l.type === 'error' || l.type === 'warning').length === 0) {
      console.log('   (none)');
    }
  } else {
    console.log('   (none)');
  }

  console.log('\n7. PAGE ERRORS:');
  if (pageErrors.length > 0) {
    pageErrors.slice(0, 10).forEach(err => {
      console.log(`   ${err.message}`);
    });
  } else {
    console.log('   (none)');
  }

  console.log('\n8. NETWORK ERRORS:');
  if (networkErrors.length > 0) {
    networkErrors.slice(0, 10).forEach(err => {
      console.log(`   ${err.url}: ${err.error}`);
    });
  } else {
    console.log('   (none)');
  }

  // Check for CSS rules that might cause thin stripes
  const cssAnalysis = await page.evaluate(() => {
    const analysis = {
      heightLimitingRules: [],
      hiddenElements: [],
      scripts: []
    };

    // Check all stylesheets
    for (const sheet of document.styleSheets) {
      try {
        for (const rule of sheet.cssRules || []) {
          if (rule.style && rule.selectorText) {
            if ((rule.selectorText.includes('epic-row') || rule.selectorText.includes('wi-container')) &&
                (rule.style.height || rule.style.minHeight || rule.style.maxHeight || rule.style.overflow)) {
              analysis.heightLimitingRules.push({
                selector: rule.selectorText,
                height: rule.style.height,
                minHeight: rule.style.minHeight,
                maxHeight: rule.style.maxHeight,
                overflow: rule.style.overflow
              });
            }
          }
        }
      } catch (e) {
        // CORS or other restrictions on stylesheet
      }
    }

    // Check for hidden/visibility issues
    const epicRows = document.querySelectorAll('.epic-row');
    epicRows.forEach((row, idx) => {
      const style = window.getComputedStyle(row);
      if (style.display === 'none' || style.visibility === 'hidden' || style.height === '0px') {
        analysis.hiddenElements.push({
          index: idx,
          reason: `display: ${style.display}, visibility: ${style.visibility}, height: ${style.height}`
        });
      }
    });

    // List scripts
    document.querySelectorAll('script').forEach((script, idx) => {
      analysis.scripts.push({
        index: idx,
        src: script.src,
        type: script.type,
        inline: !!script.textContent
      });
    });

    return analysis;
  });

  console.log('\n9. CSS/SCRIPT ANALYSIS:');
  if (cssAnalysis.heightLimitingRules.length > 0) {
    console.log('\n   Height-limiting CSS rules for .epic-row/.wi-container:');
    cssAnalysis.heightLimitingRules.forEach(rule => {
      console.log(`   - ${rule.selector}`);
      console.log(`     height: ${rule.height}, minHeight: ${rule.minHeight}, maxHeight: ${rule.maxHeight}, overflow: ${rule.overflow}`);
    });
  }

  if (cssAnalysis.hiddenElements.length > 0) {
    console.log('\n   Hidden .epic-row elements:');
    cssAnalysis.hiddenElements.forEach(elem => {
      console.log(`   - Row ${elem.index}: ${elem.reason}`);
    });
  }

  await browser.close();
  process.exit(0);
})();
